File: LeetCode/python/lc_0844.py
"""
 * @file     lc_0844.py
 * @brief    
 * @author   YongDu
 * @date     2021-10-29
"""


class Solution:
    def backspaceCompare(self, s: str, t: str) -> bool:
        def get_text(string):
            text = []
            for ch in string:
                if ch != '#':
                    text.append(ch)
                elif len(text) > 0:  # 保证栈非空
                    text.pop()
            return str(text)
        return get_text(s) == get_text(t)

File: LeetCode/cpp/lc_0844.cpp
/**
 * @file     lc_0844.cpp
 * @brief    https://leetcode-cn.com/problems/backspace-string-compare/
 * @author   YongDu
 * @date     2021-09-05
 */
class Solution {
  public:
    bool backspaceCompare(string s, string t) {
        auto getText = [](string str) {
            string lastStr;
            for (auto ch : str) {
                if (ch != '#') {
                    lastStr.push_back(ch);
                } else if (!lastStr.empty()) {
                    lastStr.pop_back();
                }
            }
            return lastStr;
        };
        return getText(s) == getText(t);
    }
};